pkg/services/sqlstore/alert_test.go

```diff
@@ -14,7 +14,7 @@ func mockTimeNow() {
 	timeNow = func() time.Time {
 		fakeNow := time.Unix(timeSeed, 0)
 		timeSeed++
-		return fakeNow
+		return fakeNow.UTC()
```

Comment:
@benpatterson Are you sure that the failure is due to time zone issues?
I'm unable to reproduce it and if I understand correctly from the description the failure is [here](https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/alert_test.go#L90).
But for the computation or both [`stateDateBeforePause`](https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/alert.go#L292) and [`stateDateAfterPause`](https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/alert.go#L354) the same function `timeNow` is used so I would expect that both would be in local timezone.

In addition to this, since the original `timeNow` [function](https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/alert.go#L14) returns local time I would prefer also the mocking function to do the same.

Could you help me reproduce this?
How do you run the tests? Do you run them all or just one of them? Do you get always this failure? Also, do you run the tests with the default database (SQLite) or any other?

thank you!

Reply:
Hi @papagian no problem!
I am on an OSX machine in the EDT timezone. I changed directory into `pkg/services/sqlstore/` and ran `go test` in order to execute tests. 

If it helps, documentation for golang points out that the `Unix` function (currently line 1148 on [this page](pkg/services/sqlstore/)), 
```
// Unix returns t as a Unix time, the number of seconds elapsed
// since January 1, 1970 UTC. The result does not depend on the
// location associated with t
```
So, I think when the mock is setting the initial time, it's using `Unix()` and setting it to 1/1/1970 and it ignores what timezone the system is in.  However, the actual alert.go function will put in a timestamp, which is location-aware. That means it sees the mocked time, looks at the local timezone, and sets the new timestamp to UTC. Hence (in my case) a little under a 5h difference.

I hope this helps!

Reply:
Thank you @benpatterson, actually it has helped a lot!
I have managed to reproduce the problem by  setting the time in my machine in EDT and in fact it can be reproduced with any timezone behind UTC.

Actually the mocking `timeNow` is using [Unix](https://golang.org/pkg/time/#Unix)  which returns:
> the local Time corresponding to the given Unix time, sec seconds and nsec nanoseconds since January 1, 1970 UTC. It is valid to pass nsec outside the range [0, 999999999]. Not all sec values have a corresponding time value. One such value is 1<<63-1 (the largest int64 value).

The problem is actually in the code not in the tests:
the `SetAlertState` is using the [Update](https://github.com/grafana/grafana/blob/master///vendor/github.com/go-xorm/xorm/session_update.go#L145:25) command to save the new state in the database while the `PauseAlert` and `PauseAllAlerts` are using [Exec](https://github.com/grafana/grafana/blob/master///vendor/github.com/go-xorm/xorm/session_raw.go#L212:25).
So eventually in the first case the local time is transformed to UTC and saved in the database but in the latter case the local time is saved in the database.

To illustrate it better, this is an alert before pause:
```
sqlite> select new_state_date from alert where id = 11;
2019-10-04 07:51:01
```

and this is the same alert after pause:

```
sqlite> select new_state_date from alert where id = 11;
2019-10-04 03:53:16.962685-04:00
```

So instead of changing the mocking class to return UTC the [`PauseAlert`](https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/alert.go#L320) and [`PauseAllAlerts`](url) should be modified in order to pass UTC to the `Exec`.
Would you like to perform these modifications?

Thanks again for reporting this!